client/asset/eth/eth.go

```diff
@@ -4594,177 +4452,592 @@ func (w *baseWallet) emitTransactionNote(tx *asset.WalletTransaction, new bool)
+		if err != nil {
+			w.log.Errorf("Error getting header for hash %v: %v", receipt.BlockHash, err)
+			return
+		}
+		if hdr == nil {
+			w.log.Errorf("Header for hash %v is nil", receipt.BlockHash)
 			return
 		}
+		pendingTx.Timestamp = hdr.Time
+		if effectiveGasPrice == nil {
+			effectiveGasPrice = new(big.Int).Add(hdr.BaseFee, tx.EffectiveGasTipValue(hdr.BaseFee))
+		}
+	}
 
-		if updated || !pendingTx.savedToDB {
-			err := w.txDB.storeTx(nonce, pendingTx)
-			if err != nil {
-				w.log.Errorf("error updating tx in db: %w", err)
-				pendingTx.savedToDB = false
-				return
-			}
+	bigFees := new(big.Int).Mul(effectiveGasPrice, big.NewInt(int64(receipt.GasUsed)))
+	pendingTx.Fees = dexeth.WeiToGweiCeil(bigFees)
+	pendingTx.BlockNumber = receipt.BlockNumber.Uint64()
+	pendingTx.Confirmed = safeConfs(tip, pendingTx.BlockNumber) >= w.finalizeConfs
+}
 
-			pendingTx.savedToDB = true
-			if pendingTx.Confirmed {
-				w.pendingTxsMtx.Lock()
-				delete(w.pendingTxs, nonce)
-				w.pendingTxsMtx.Unlock()
-			}
+// checkPendingTxs checks the confirmation status of all pending transactions.
+func (w *baseWallet) checkPendingTxs() {
+	tip := w.tipHeight()
 
-			w.emitTransactionNote(pendingTx.WalletTransaction, false)
+	w.nonceMtx.Lock()
+	defer w.nonceMtx.Unlock()
+
+	// If we have pending txs, trace log the before and after.
+	if nPending := len(w.pendingTxs); nPending > 0 && w.log.Level() == dex.LevelTrace {
+		defer func() {
+			w.log.Tracef("Checked %d pending txs. Finalized %d", nPending, nPending-len(w.pendingTxs))
+		}()
+	}
+
+	// keepFromIndex will be the index of the first un-finalized tx.
+	// disorderedConfsIndex, if non-zero, will be the index of the last
+	// confirmed tx after the first unconfirmed tx, if one exists. It is used to
+	// detect an abnormal condition where the tx was nonce-replaced and we don't
+	// know the replacement tx.
+	var keepFromIndex, lastConfirmed int
+	for i, pendingTx := range w.pendingTxs {
+		if w.ctx.Err() != nil {
+			return
 		}
-	}()
+		w.updatePendingTx(tip, pendingTx)
+		if pendingTx.Confirmed {
+			lastConfirmed = i
+			if pendingTx.Nonce.Cmp(w.confirmedNonceAt) == 0 {
+				w.confirmedNonceAt.Add(w.confirmedNonceAt, big.NewInt(1))
+			}
+			if pendingTx.savedToDB {
+				if i == keepFromIndex {
+					// This is what we expect. No tx should be confirmed before a
+					// tx with a lower nonce. We'll delete at least up to this
+					// one.
+					keepFromIndex = i + 1
+				}
+			}
+			// This transaction is finalized. If we had previously sought action
+			// on it, cancel that request.
+			if pendingTx.actionRequested {
+				pendingTx.actionRequested = false
+				w.requestAction(asset.ActionResolved, pendingTx.ID, nil, pendingTx.TokenID)
+			}
+		}
+	}
 
-	if pendingTx.lastCheck == tip {
-		return false
+	// If we have missing nonces, send an alert.
+	if !w.recoveryRequestSent && len(findMissingNonces(w.confirmedNonceAt, w.pendingNonceAt, w.pendingTxs)) != 0 {
+		w.recoveryRequestSent = true
```

Comment:
There's a small note about it in `userActionRecoverNonces`.
```
// Don't reset recoveryRequestSent. They won't see this message again until they reboot.
```
My fear is that a user runs two instances at once and somehow gets a bunch of these and they just keep spamming the button.

Reply:
I guess it's find to have them reboot in case this happens. I tested the missing nonces, accepted, and then it did the replacement transactions. When I refreshed the page it showed the popup to handle missing nonces again.